ppp/discovery.go

```diff
@@ -14,40 +20,303 @@ const (
+)
+
+// Constants for PPPoE Discovery tag types
+const (
+	pppoeTagServiceName = 0x0101 // Roughly speaking, the name of the ISP.
+	pppoeTagACName      = 0x0102 // Roughly speaking, the hostname of the PPPoE concentrator.
+	pppoeTagCookie      = 0x0104 // The PPPoE equivalent of a syncookie.
+)
+
 var (
 	// padiPacket is a PPPoE Active Discovery Initiation (PADI) packet
 	// that sollicits session offers from any available PPPoE
 	// concentrator.
-	padiPacket = []byte{
-		0x11, // Protocol version 1, packet type 1 (both constants)
-		9,    // Code PADI, aka "hey any PPPoE concentrators out there?"
-		0, 0, // Session ID, not set during discovery
-		0, 4, // Payload length. payload is a TLV array of tags, but we're sending only 1 zero-length tag
-		1, 1, // Tag "Service-Name", aka "which ISP is acceptable to me?
-		0, 0, // Tag value length, 0
-		// No tag value, meaning "any ISP is acceptable"
-	}
+	padiPacket = encodeDiscoveryPacket(&discoveryPacket{
+		Code: pppoePADI,
+		TLV: map[int][]byte{
+			// By convention on single-ISP customer access networks,
+			// the tag is always nil, meaning "don't care," because
+			// there's only one ISP around anyway.
+			pppoeTagServiceName: nil,
+		},
+	})
 	// ethernetBroadcast is the Ethernet broadcast address.
 	ethernetBroadcast = &raw.Addr{
 		HardwareAddr: net.HardwareAddr{0xff, 0xff, 0xff, 0xff, 0xff, 0xff},
 	}
 )
 
 // pppoeDiscovery executes PPPoE discovery and returns a PPPoE session ID.
-func pppoeDiscovery(ifName string) (sessionID uint16, err error) {
+func pppoeDiscovery(ctx context.Context, ifName string) (sessionID int, closer func() error, err error) {
+	conn, err := newDiscoveryConn(ifName)
+	if err != nil {
+		return 0, nil, err
+	}
+	defer func() {
+		if err != nil {
+			conn.Close()
+		}
+	}()
+
+	deadline, hasDeadline := ctx.Deadline()
+
+	var (
+		concentrator net.Addr
+		cookie       []byte
+	)
+
+	// Broadcast PADIs, looking for a PPPoE concentrator.
+	for concentrator == nil && (!hasDeadline || time.Now().Before(deadline)) {
+		// Send a PADI, asking concentrators for a session offer.
+		if err := sendPADI(conn); err != nil {
+			return 0, nil, fmt.Errorf("sending PADI packet: %v", err)
+		}
+
+		padoCtx, cancelPADO := context.WithTimeout(ctx, time.Second)
+		defer cancelPADO()
+		concentrator, cookie, err = readPADO(padoCtx, conn)
+		if err == nil {
+			// We know about a concentrator, move on.
+			break
+		} else if neterr, ok := err.(net.Error); !ok || !neterr.Timeout() {
+			return 0, nil, fmt.Errorf("waiting for PADO: %v", err)
+		}
+		// Timed out waiting for PADO. Loop back around to (maybe) try
+		// again.
+	}
+
+	// Got a concentrator, request a session.
+	for !hasDeadline || time.Now().Before(deadline) {
+		if err := sendPADR(conn, concentrator, cookie); err != nil {
+			return 0, nil, fmt.Errorf("sending PADR packet: %v", err)
+		}
+
+		padsCtx, cancelPADS := context.WithTimeout(ctx, time.Second)
+		defer cancelPADS()
```

Comment:
defer is func-scoped, not block-scoped, but that works fine here: you only want to cancel these quickly on success (where you already return quickly). On failure they've already fired so it's not necessary to clean up quickly.

Reply:
Yeah, defers may accumulate here if the caller specifies an extremely long deadline (thousands of seconds). Something to clean up along with other API changes.